[Tooling] Fix `finalize_release` and the `release/*` branch checkout

AliSoftware opened the pull request:
# Why?

During release process we need the lanes invoked from the `release-pipelines/*.yml` and that will require to do a `git push` to the release branch to actually `git fetch` + `git checkout` those release branches explicitly, otherwise the `git push` will fail (e.g. unknown `refspec`)

Example of a failure of such case: https://buildkite.com/automattic/pocket-casts-ios/builds/9594#019504fc-44dd-433a-a9fe-39cf81f67510

# How

 - We still need to checkout the `release/*` branch before the compilation in `release-build.sh`, because we want to ensure that we include the latest commit that was just pushed by the `new_beta_release` / `finalize_release` lanes from previous job, instead of building the commit that the CI build was initially created on
 - But we **also** need to checkout the `release/*` in `release-pipelines/*.yml` before calling any of the lanes that would do a version bump and push the commit, otherwise it won't try to do the commit on the branch but on the detached HEAD.

# Links

Internal ref: p1739547154659529/1739543498.075849-slack-CC7L49W13

AliSoftware committed:
Fix finalize-release and release branch checkout

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/Automattic/pocket-casts-ios/milestone/77">7.82 ❄️</a>. The due date for this milestone has already passed.<br />
Please assign it to a milestone with a later deadline or check whether the release for this milestone has already been finished.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

AliSoftware committed:
Still checkout release branch before doing build

To be sure to include the new commit made by the other job instead of building from the previous commit the CI build started on

iangmaia commented on .buildkite/release-pipelines/finalize-hotfix-release.yml:
```diff
@@ -18,6 +18,8 @@ steps:
       echo '--- 🔐 Access Secrets'
       bundle exec fastlane run configure_apply
 
+      .buildkite/commands/checkout-release-branch.sh "${RELEASE_VERSION}"
```
Shouldn't the checkout happen before `install_gems`, just to be 100% consistent we're going to install the exact gems that are in the branch? (this also applies to the other pipelines ofc).

AliSoftware replied:
Good catch! Indeed.

iangmaia committed:
Move checkout-release-branch.sh call before doing anything in the branch
